### PR TITLE
feat:add snap build

### PR DESCRIPTION
**Pull request by skourta**
This pull request includes the addition of a new `snapcraft.yaml` file to configure the packaging of the `charmed-etcd-snap` application. The most important changes involve defining the snap package metadata, specifying the applications included in the snap, and detailing the parts and build instructions.

Snap package configuration:

* Added `name`, `base`, `version`, `summary`, and `description` fields to define the snap package metadata.
* Set `grade` to `devel` and `confinement` to `strict` for the snap package.

Application definitions:

* Defined three applications (`etcd`, `etcdctl`, and `etcdutl`) with their respective commands and required plugs.

Build instructions:

* Specified the `etcd` part with `make` plugin, build dependencies, source repository, and custom build commands.

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      - network-bind
+      - home
+
+parts:
+  etcd:
+    plugin: make
+    build-snaps: [go/latest/stable]
+    source: https://git.launchpad.net/etcd
+    source-type: git
+    source-branch: main
```
Comment (reneradoi) [inline snap/snapcraft.yaml]: This should be changed to the 3.5 branch in Launchpad.

**Review by reneradoi — COMMENTED** [review]
Hey Smail, thanks for the PR!

What still needs to be added is the environment variables:
- Those defining e.g. the default $SNAP-directories can be set as global environment as in [this example](https://github.com/canonical/opensearch-snap/blob/2/edge/snap/snapcraft.yaml#L57).
- As a specific environment for the etcd snap to make etcd configurable we need to set `ETCD_CONFIG_FILE` to the config path, just like in [this example](https://github.com/canonical/opensearch-snap/blob/2/edge/snap/snapcraft.yaml#L103C5-L104C59).

**Review by reneradoi — APPROVED** [review]
Thanks for the changes!

File: snap/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    plugs:
+      - network-bind
+    environment:
+      ETCD_CONFIG_FILE: $SNAP_CURRENT/etcd.conf.yml
+
+  etcd:
+    command: bin/etcd
+    plugs:
+      - network-bind
+      - home
```
Comment (Mehdi-Bendriss) [inline snap/snapcraft.yaml:29]: Why is this plug needed? this should give access to the home directory of the user.

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  etcdctl:
+    command: bin/etcdctl
+    plugs:
+      - network-bind
+      - home
```
Comment (Mehdi-Bendriss) [inline snap/snapcraft.yaml:34]: Similar comment as above

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  etcdutl:
+    command: bin/etcdutl
+    plugs:
+      - network-bind
+      - home
```
Comment (Mehdi-Bendriss) [inline snap/snapcraft.yaml:39]: similar comment as above

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+parts:
+  etcd:
+    plugin: make
+    build-snaps: [go/latest/stable]
+    source: https://git.launchpad.net/etcd
+    source-type: git
+    source-branch: lp-v3.5.16
```
Comment (Mehdi-Bendriss) [inline snap/snapcraft.yaml:47]: shouldn't this be `source-tag` ?

File: etcd.conf.yml
Comment (Mehdi-Bendriss) [inline etcd.conf.yml:1]: This should be put in a directory `config` under `snap`

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    override-build: |
+      make build
+      cp -r bin $CRAFT_PART_INSTALL/
+  
+  etcd-conf-file:
+    plugin: dump
+    source: .
+    stage:
+      - etcd.conf.yml
```
Comment (Mehdi-Bendriss) [inline snap/snapcraft.yaml:56]: nit:
```suggestion
      - etcd.conf.yml

```

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  SNAP_DATA_CURRENT: /var/snap/charmed-etcd/current
+
+apps:
+  daemon:
+    daemon: simple
+    command: bin/etcd
+    plugs:
+      - network-bind
+    environment:
+      ETCD_CONFIG_FILE: $SNAP_CURRENT/etcd.conf.yml
```
Comment (Mehdi-Bendriss) [inline snap/snapcraft.yaml:23]: We should refer to this file from `$SNAP_DATA_CURRENT` -- as `$SNAP_CURRENT` is a read only directory.
```suggestion
      ETCD_CONFIG_FILE: $SNAP_DATA_CURRENT/etcd.conf.yml
```

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Smail, great work! I left some comments.
Please also put the licenses in a `licenses` directory as done [here](https://github.com/canonical/opensearch-snap/tree/2/edge/licenses). Where 2 license files should be copied: `LICENSE-etcd` and `LICENSE-snap`.

File: snap/snapcraft.yaml (hunk 2 above)
Comment (skourta) [reply snap/snapcraft.yaml:29]: To enable users to run etcd with a data folder located in the home directory. I added the etcd command to enable users to run any instance of etcd they want.

File: snap/snapcraft.yaml (hunk 3 above)
Comment (skourta) [reply snap/snapcraft.yaml:34]: We need the home plug if a user needs to make a snapshot and save it in their home. If this use case is not supported then we can remove it.

File: snap/snapcraft.yaml (hunk 4 above)
Comment (skourta) [reply snap/snapcraft.yaml:39]: Restoring snapshots from the home directory will need access their.

**Comment by skourta** [comment]
> Thanks Smail, great work! I left some comments. Please also put the licenses in a `licenses` directory as done [here](https://github.com/canonical/opensearch-snap/tree/2/edge/licenses). Where 2 license files should be copied: `LICENSE-etcd` and `LICENSE-snap`.

Will do. I followed the [PgBouncer](https://github.com/canonical/charmed-pgbouncer-snap) snap as it was the most recent charm I could find.

File: snap/snapcraft.yaml (hunk 2 above)
Comment (Mehdi-Bendriss) [reply snap/snapcraft.yaml:29]: This should not be necessary. The data folder should be located in `$SNAP_COMMON/var/data` and the logs folder should be in `$SNAP_COMMON/var/log` 
Please refer to the OpenSearch snap for the paths.

**Pull request closed without merging**